app/controllers/merchants_controller.rb

```diff
@@ -0,0 +1,43 @@
+class MerchantsController < ApplicationController
+
+  def index
+    @merchants = Merchant.all
+  end
+
+  def show
+    @merchant = Merchant.find_by(id: params[:id])
+    if @merchant.nil?
+      flash[:warning] = "Merchant with id #{params[:id]} was not found."
+      redirect_to root_path
+    end
+  end
+
+  def create
+    auth_hash = request.env["omniauth.auth"]
+    merchant = Merchant.find_by(uid: auth_hash[:uid], provider: "github")
+
+    if merchant
+      flash[:success] = "Logged in as returning merchant #{merchant.username}."
+    else
+      merchant = Merchant.build_from_github(auth_hash)
+
+      if merchant.save
+        flash[:success] = "Logged in as new merchant #{merchant.username}."
+      else
+        flash[:error] = "Could not create new merchant account: #{merchant.errors.messages}"
+        return redirect_to root_path
+      end 
+    end
+
```

Comment:
Consider adding the current_user controller action to track the presence of the current session user id.

Reply:
@current_user = User.find_by(id: session[:user_id])
    unless @current_user
      flash[:error] = "You must be logged in to see this page"
      redirect_to root_path
    end
  end